File: src/indicators.ts
import {
  RSI,
  BollingerBands,
  MACD,
} from 'technicalindicators';
import { Bar, IndicatorResult } from './types';

// ─── RSI (14) ──────────────────────────────────────────────────────────────

export function computeRSI(bars: Bar[], period: number = 14): number[] {
  const closes = bars.map((b) => b.close);
  return RSI.calculate({ values: closes, period });
}

// ─── Bollinger Bands (20, 2) ────────────────────────────────────────────────

export function computeBB(
  bars: Bar[],
  period: number = 20,
  stdDev: number = 2,
): { upper: number[]; middle: number[]; lower: number[] } {
  const closes = bars.map((b) => b.close);
  const results = BollingerBands.calculate({ values: closes, period, stdDev });
  return {
    upper: results.map((r) => r.upper),
    middle: results.map((r) => r.middle),
    lower: results.map((r) => r.lower),
  };
}

// ─── MACD (12, 26, 9) ──────────────────────────────────────────────────────

export function computeMACD(
  bars: Bar[],
  fastPeriod: number = 12,
  slowPeriod: number = 26,
  signalPeriod: number = 9,
): { MACD: number[]; signal: number[]; histogram: number[] } {
  const closes = bars.map((b) => b.close);
  const results = MACD.calculate({
    values: closes,
    fastPeriod,
    slowPeriod,
    signalPeriod,
    SimpleMAOscillator: false,
    SimpleMASignal: false,
  });
  return {
    MACD: results.map((r) => r.MACD ?? 0),
    signal: results.map((r) => r.signal ?? 0),
    histogram: results.map((r) => r.histogram ?? 0),
  };
}

// ─── All indicators at once ─────────────────────────────────────────────────

export function computeIndicators(bars: Bar[]): IndicatorResult {
  return {
    rsi: computeRSI(bars),
    bb: computeBB(bars),
    macd: computeMACD(bars),
  };
}

// ─── Candle pattern helpers ─────────────────────────────────────────────────

/**
 * Returns true when `curr` is a bullish engulfing candle vs `prev`.
 * Bullish engulfing: prev is bearish, curr is bullish, curr body wraps prev body.
 */
export function isBullishEngulfing(prev: Bar, curr: Bar): boolean {
  const prevBearish = prev.close < prev.open;
  const currBullish = curr.close > curr.open;
  const wraps = curr.open <= prev.close && curr.close >= prev.open;
  return prevBearish && currBullish && wraps;
}

/**
 * Returns true when `curr` is a bearish engulfing candle vs `prev`.
 * Bearish engulfing: prev is bullish, curr is bearish, curr body wraps prev body.
 */
export function isBearishEngulfing(prev: Bar, curr: Bar): boolean {
  const prevBullish = prev.close > prev.open;
  const currBearish = curr.close < curr.open;
  const wraps = curr.open >= prev.close && curr.close <= prev.open;
  return prevBullish && currBearish && wraps;
}

File: indicators.py
"""
indicators.py — RSI(14), Bollinger Bands(20,2), MACD(12,26,9), Volume Avg(20), VWAP
              + candlestick patterns.
Pure pandas/numpy implementation — no external indicator library needed.
"""

import numpy as np
import pandas as pd


def _rsi(series: pd.Series, period: int = 14) -> pd.Series:
    """Compute RSI using exponential moving average of gains/losses."""
    delta = series.diff()
    gain = delta.where(delta > 0, 0.0)
    loss = -delta.where(delta < 0, 0.0)
    avg_gain = gain.ewm(alpha=1 / period, min_periods=period).mean()
    avg_loss = loss.ewm(alpha=1 / period, min_periods=period).mean()
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def _bbands(series: pd.Series, period: int = 20, std: float = 2.0) -> tuple:
    """Compute Bollinger Bands (upper, middle, lower)."""
    middle = series.rolling(window=period).mean()
    rolling_std = series.rolling(window=period).std()
    upper = middle + std * rolling_std
    lower = middle - std * rolling_std
    return upper, middle, lower


def _macd(series: pd.Series, fast: int = 12, slow: int = 26, signal: int = 9) -> tuple:
    """Compute MACD line, signal line, and histogram."""
    ema_fast = series.ewm(span=fast, adjust=False).mean()
    ema_slow = series.ewm(span=slow, adjust=False).mean()
    macd_line = ema_fast - ema_slow
    signal_line = macd_line.ewm(span=signal, adjust=False).mean()
    histogram = macd_line - signal_line
    return macd_line, signal_line, histogram


def _vwap(df: pd.DataFrame) -> pd.Series:
    """
    Compute VWAP, reset each trading day.
    VWAP = cumsum(typical_price × volume) / cumsum(volume) per day.
    Requires columns: high, low, close, volume, time.
    """
    typical_price = (df["high"] + df["low"] + df["close"]) / 3
    tp_vol = typical_price * df["volume"]

    # Group by calendar date so VWAP resets at midnight
    dates = pd.to_datetime(df["time"]).dt.date

    df_temp = df.assign(_tp_vol=tp_vol, _vol=df["volume"], _date=dates)
    cum_tp_vol = df_temp.groupby("_date")["_tp_vol"].cumsum()
    cum_vol    = df_temp.groupby("_date")["_vol"].cumsum()

    return cum_tp_vol / cum_vol


def compute_indicators(df: pd.DataFrame) -> pd.DataFrame:
    """Compute RSI(14), Bollinger Bands(20,2), MACD(12,26,9), Volume Avg(20), and VWAP."""
    df = df.copy()

    closes  = df["close"]
    volumes = df["volume"]

    # RSI
    df["RSI_14"] = _rsi(closes, 14)

    # Bollinger Bands
    df["BBU_20_2.0"], df["BBM_20_2.0"], df["BBL_20_2.0"] = _bbands(closes, 20, 2.0)

    # MACD
    df["MACD_12_26_9"], df["MACDs_12_26_9"], df["MACDh_12_26_9"] = _macd(closes, 12, 26, 9)

    # Volume climax filter: 20-bar average volume
    df["volume_avg_20"] = volumes.rolling(window=20).mean()

    # VWAP (resets each trading day)
    df["VWAP"] = _vwap(df)

    return df


def is_bullish_engulfing(prev: pd.Series, curr: pd.Series) -> bool:
    """Bullish engulfing: prev bearish, curr bullish, curr body wraps prev body."""
    prev_bearish = prev["close"] < prev["open"]
    curr_bullish = curr["close"] > curr["open"]
    wraps = curr["open"] <= prev["close"] and curr["close"] >= prev["open"]
    return prev_bearish and curr_bullish and wraps


def is_bearish_engulfing(prev: pd.Series, curr: pd.Series) -> bool:
    """Bearish engulfing: prev bullish, curr bearish, curr body wraps prev body."""
    prev_bullish = prev["close"] > prev["open"]
    curr_bearish = curr["close"] < curr["open"]
    wraps = curr["open"] >= prev["close"] and curr["close"] <= prev["open"]
    return prev_bullish and curr_bearish and wraps
